python/tutorials/09-persistent-matmul.py

```diff
@@ -359,6 +359,127 @@ def matmul_tma_persistent(a, b):
     return c
 
 
+@triton.jit(launch_metadata=_matmul_launch_metadata)
+def matmul_kernel_device_tma_persistent(a_desc_ptr, b_desc_ptr, c_desc_ptr,  #
+                                        a_ptr, b_ptr, c_ptr,  #
+                                        ready_flag,  #
+                                        M, N, K,  #
+                                        BLOCK_SIZE_M: tl.constexpr,  #
+                                        BLOCK_SIZE_N: tl.constexpr,  #
+                                        BLOCK_SIZE_K: tl.constexpr,  #
+                                        GROUP_SIZE_M: tl.constexpr,  #
+                                        NUM_SMS: tl.constexpr):  #
+    # Matmul using TMA and device-side descriptor creation
+    dtype = c_ptr.dtype.element_ty
+    start_pid = tl.program_id(axis=0)
+    num_pid_m = tl.cdiv(M, BLOCK_SIZE_M)
+    num_pid_n = tl.cdiv(N, BLOCK_SIZE_N)
+    k_tiles = tl.cdiv(K, BLOCK_SIZE_K)
+    num_tiles = num_pid_m * num_pid_n
+
+    if start_pid == 0:
+        tl.extra.cuda.experimental_device_tensormap_create2d(desc_ptr=a_desc_ptr, global_address=a_ptr,
+                                                             load_size=[BLOCK_SIZE_M, BLOCK_SIZE_K], global_size=[M, K],
+                                                             element_ty=a_ptr.dtype.element_ty)
+        tl.extra.cuda.experimental_device_tensormap_create2d(desc_ptr=b_desc_ptr, global_address=b_ptr,
+                                                             load_size=[BLOCK_SIZE_N, BLOCK_SIZE_K], global_size=[N, K],
+                                                             element_ty=b_ptr.dtype.element_ty)
+        tl.extra.cuda.experimental_device_tensormap_create2d(desc_ptr=c_desc_ptr, global_address=c_ptr,
+                                                             load_size=[BLOCK_SIZE_M, BLOCK_SIZE_N], global_size=[M, N],
+                                                             element_ty=c_ptr.dtype.element_ty)
+        tl.atomic_xchg(ready_flag, 1, sem="release")
+    else:
+        flag = tl.full([], 0, tl.int32)
+        while flag != 1:
+            flag = tl.atomic_add(ready_flag, 0, sem="acquire")
```

Comment:
Could (and should?) we use tl.debug_barrier here?

Reply:
I don't think it is needed since we check the result of the atomic?

Reply:
ah, the point is to sync across blocks here so that wouldn't be enough